keps/sig-node/5055-device-taints-and-tolerations/README.md

```diff
@@ -0,0 +1,971 @@
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+The feature is following the approach and APIs taken for node taints and
+applies them to devices. A new controller watches tainted devices and deletes
+pods using them unless they tolerate the device taint, similar to the
+[taint-eviction-controller](https://github.com/kubernetes/kubernetes/blob/32130691a4cb8a1034b999341c40e48d197f5465/pkg/controller/tainteviction/taint_eviction.go#L81-L83).
+
+Taints are cumulative as long as the key and effect pairs are different:
+- Taints defined by an admin in a ResourceSliceOverride get added to the
+  set of taints defined by the DRA driver in a ResourceSlice.
+- Taints with the same key and effect get overwritten, using the same
+  precedence as for attributes.
+
+This merging will be implemented by the same code that also
+overrides device attributes.
+
+To ensure consistency among all pods sharing a ResourceClaim, the toleration
+for taints gets added to the request in a ResourceClaim, not the pod. This also
+avoids conflicts like one pod tolerating a taint for scheduling and some other
+pod not tolerating that.
+
+### API
+
+The ResourceSlice content gets extended:
+
+```Go
+// BasicDevice defines one device instance.
+type BasicDevice struct {
+    ...
+
+    // If specified, the device's taints.
+    //
+    // This is an alpha field and requires enabling the DRADeviceTaints
+    // feature gate.
+    //
+    // +optional
+    // +listType=atomic
+    // +featureGate=DRADeviceTaints
+    Taints []Taint
+}
+
+// The device this Taint is attached to has the "effect" on
+// any claim and, through the claim, to pods that do not tolerate
+// the Taint.
+type Taint struct {
+    // The taint key to be applied to a device.
+    // Must be a label name.
+    //
+    // +required
+    Key string
+
+    // The taint value corresponding to the taint key.
+    // Must be a label value.
+    //
+    // +optional
+    Value string
+
+    // The effect of the taint on claims and pods that do not tolerate the taint.
+    // Valid effects are NoSchedule and NoExecute. PreferNoSchedule as used for
+    // nodes is not valid here.
+    //
+    // +required
+    Effect TaintEffect
+
+    // ^^^^
+    //
+    // Implementing PreferNoSchedule would depend on a scoring solution for DRA.
+    // It might get added as part of that.
+
+    // TimeAdded represents the time at which the taint was added.
+    // It is only written for NoExecute taints.
+    //
+    // +optional
+    TimeAdded *metav1.Time
+}
+```
+
+Taint has the exact same fields as a v1.Taint, but the description is a bit
+different. In particular, PreferNoSchedule is not valid.
```

Comment:
Are there any strong reason for "PreferNoSchedule" is not valid"?? I supposed we can support this _after_ we supported device scoring(https://github.com/kubernetes/enhancements/issues/4970). If the reason was device scoring is not released yet, could we support this in the future?

Reply:
The API must only allow what is actually supported. If PreferNoSchedule is not implemented, then it should get rejected.

Yes, this support can be added later, at which point we need to relax the validation and allow that value.